Add DifferenceFromAnalyticCompute tag

geoffrey4444 opened the pull request:
## Proposed changes

Add a compute tag `DifferenceFromAnalyticCompute` that takes a `Variables` from a `DataBox`, subtracts the analytic solution `Analytic<Tag>` from each `Tag` in that `Variables`, and puts the result in the `DataBox` in a `Variables` whose tags are wrapped in the new `Difference<>` prefix tag.

This is a prerequisite to modifying `ObserveErrorNorms` so that we can output other error measures, such as constraints; currently, `ObserveErrorNorms` can only output differences from the analytic solution.

I also updated `WrappedGr`, so that it has the same `tags` alias as all other `AnalyticSolution`s do, so that it can work properly with `DifferenceFromAnalyticCompute`, and I moved `Tags::Analytic` to `PointwiseFunctions/AnalyticSolutions/Tags.hpp`.

Thanks to @moxcodes @markscheel for helping me out a lot in understanding how the template meta-programming works in this change!

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Move Analytic<> tag to AnalyticSolutions

geoffrey4444 committed:
Add tag to compute diff from analytic solution

geoffrey4444 committed:
Add tags alias to WrappedGr

kidder commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -10,6 +10,7 @@
 #include "DataStructures/Variables.hpp"
 #include "DataStructures/VariablesTag.hpp"
 #include "Domain/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
```
remove the include for Prefixes

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -3,6 +3,7 @@
 
 #pragma once
 
+#include "DataStructures/DataBox/DataBoxTag.hpp"
```
do not add

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -73,4 +74,14 @@ struct BoundaryCondition : BoundaryConditionBase, db::SimpleTag {
+/// \snippet Test_DataBoxPrefixes.cpp analytic_name
+template <typename Tag>
+struct Analytic : db::PrefixTag, db::SimpleTag {
+  using type = db::const_item_type<Tag>;
```
just use `Tag::type`

kidder commented on tests/Unit/Elliptic/Actions/Test_InitializeAnalyticSolution.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "Elliptic/Actions/InitializeAnalyticSolution.hpp"
 #include "Framework/ActionTesting.hpp"
 #include "Parallel/ParallelComponentHelpers.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
```
remove the prefix include

kidder commented on tests/Unit/Evolution/CMakeLists.txt:
```diff
@@ -11,7 +11,6 @@ add_subdirectory(VariableFixing)
 set(LIBRARY "Test_Evolution")
 
 set(LIBRARY_SOURCES
-  Test_ComputeTags.cpp
```
why is this removed?

geoffrey4444 replied:
Thanks for catching this...I didn't mean to remove it

kidder commented on tests/Unit/Evolution/Test_ComputeTags.cpp:
```diff
@@ -14,6 +14,7 @@
 #include "DataStructures/Variables.hpp"
 #include "Evolution/ComputeTags.hpp"
 #include "Helpers/DataStructures/DataBox/TestHelpers.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
```
remove the prefix include

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+ * solution.
+ */
+template <typename Tag>
+struct Difference : db::PrefixTag, db::SimpleTag {
```
Call it Error not Difference; what other difference would there be?  I would have no idea of what, for example,`Difference<Lapse>` would mean...

geoffrey4444 replied:
I was struggling for a good name...I like this suggestion!

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+      db::ComputeTag {
+  using argument_tags = tmpl::list<
+      Tags::Variables<TagsList>,
+      db::add_tag_prefix<Tags::Analytic, Tags::Variables<AnalyticTagsList>>>;
```
why isn't this simply `Tags::Variables<AnalyticTagsList>` ?

geoffrey4444 replied:
The analytic Variables is keyed by Tags::Analytic<Tag> , while the numeric Variables is keyed by Tag. I couldn't figure out how to do what you suggest, because then how do I figure out which Tag goes with which AnalyticTag? But if there's a better way, Id welcome the suggestion!

kidder replied:
where does `AnalyticTagsList` come from?  Why isn't it `tmpl::list<Tags::Analytic<Var0>,...Tags::Analytic<VarN>>` ?

nilsvu replied:
@kidder For variables tags we need the `add_tag_prefix` because both the variables tag as well as its containing tags must be prefixed to work correctly.

And I believe the list of tags we have an analytic solution for can be a subset of the full list of system variables, so I think it makes sense to have the separate template parameter. It could default to the full tags list though.

How about you replace the `TagsList` template parameter with `VariablesTag`? You can get its tags list as `db::get_variables_tags_list<VariablesTag>` so you have the same information available, but you don't have to build the `::Tags::Variables`. If the tags in `TagsList` carry prefixes then the code you currently have would not compile.

geoffrey4444 replied:
After working on this some more, I figured out what's going on here.

Yes, AnalyticTagsList is `temple::list<Tags::Analytic<Varo>, ... Tags::Analytic<VarN>>`.

The trick I didn't see before was that it should actually be

`Tags::Analytic<Tags::Variables<AnalyticTagsList>>`, because not only are the tags in the `Variables` wrapped in `Tags::Analytic`, but the `Variables` itself is as well.

This change comes at the price that users that call `DifferenceFromAnalyticCompute` must themselves wrap the list of tags provided by the analytic solution in Tags::Analytic<>, because the solutions' `tags` aliases don't do this.

I figure you would be ok with that price, so I made this change as you requested and updated the test that calls it accordingly.

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+  static constexpr void function(
+      const gsl::not_null<return_type*> error,
+      const ::Variables<TagsList>& numeric,
+      const ::Variables<db::wrap_tags_in<Tags::Analytic, AnalyticTagsList>>&
```
again why isn't this `::Variables<AnalyticTagsList>` ?

geoffrey4444 replied:
The compute tag for the analytic solution wraps each key in Tags::Analytic<>

geoffrey4444 replied:
See above: I figured out how to make this change. I had to change the argument tag to `Tags::Analytic<Tags::Variables<AnalyticTagsList>>`, not just `Tags::Variables<AnalyticTagsList>`, and I had to change the test to pass `db::wrap_tags_in<Tags::Analytic, solution::tags<DataVector>>` for the second template parameter of `Tags::DifferenceFromAnalyticCompute` instead of simply passing `solution::tags<DataVector>` as before

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+      const ::Variables<TagsList>& numeric,
+      const ::Variables<db::wrap_tags_in<Tags::Analytic, AnalyticTagsList>>&
+          analytic) noexcept {
+    (*error) = return_type{numeric.number_of_grid_points()};
```
how about just `*error = numeric` ?

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+    tmpl::for_each<TagsList>([&error, &numeric, &analytic](auto tag_v) {
+      using tensor_tag = tmpl::type_from<decltype(tag_v)>;
+      for (size_t i = 0; i < db::item_type<tensor_tag>::size(); ++i) {
+        (get<Tags::Difference<tensor_tag>>(*error))[i] =
```
then this just becomes a `-=` of the analytic solution

kidder commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+/*!
+ * \brief For each `Tag` in `TagsList`, where `TagsList` is a subset of
+ * `AnalyticTagsList`, compute the difference `Tag` - `Tags::Analytic<Tag>`.
+ */
```
is the compiler error comprehensible if `TagsList` is not a subset of `AnalyticTagsList` ?

geoffrey4444 replied:
I think so. If you remove WrappedGr from the solution in the test (so SpacetimeMetric, Phi, and Pi, the three tags in TagsList, are not in the analytic solution), then you get this error, which says "can't get SpacetimeMetric from Variables" and points to the precise line of Errors.hpp where you try to get that tag out of the analytic solution

error: static_assert failed due to requirement 'tmpl::list_contains_v<brigand::list<Tags::Analytic<gr::Tags::Lapse<DataVector> >, Tags::Analytic<Tags::dt<gr::Tags::Lapse<DataVector> > >, Tags::Analytic<Tags::deriv<gr::Tags::Lapse<DataVector>, brigand::integral_constant<unsigned long, 1>, Frame::Inertial, nullptr_t> >, Tags::Analytic<gr::Tags::Shift<1, Frame::Inertial, DataVector> >, Tags::Analytic<Tags::dt<gr::Tags::Shift<1, Frame::Inertial, DataVector> > >, Tags::Analytic<Tags::deriv<gr::Tags::Shift<1, Frame::Inertial, DataVector>, brigand::integral_constant<unsigned long, 1>, Frame::Inertial, nullptr_t> >, Tags::Analytic<gr::Tags::SpatialMetric<1, Frame::Inertial, DataVector> >, Tags::Analytic<Tags::dt<gr::Tags::SpatialMetric<1, Frame::Inertial, DataVector> > >, Tags::Analytic<Tags::deriv<gr::Tags::SpatialMetric<1, Frame::Inertial, DataVector>, brigand::integral_constant<unsigned long, 1>, Frame::Inertial, nullptr_t> >, Tags::Analytic<gr::Tags::SqrtDetSpatialMetric<DataVector> >, Tags::Analytic<gr::Tags::ExtrinsicCurvature<1, Frame::Inertial, DataVector> >, Tags::Analytic<gr::Tags::InverseSpatialMetric<1, Frame::Inertial, DataVector> > >, Tags::Analytic<gr::Tags::SpacetimeMetric<1, Frame::Inertial, DataVector> > >' "Could not retrieve Tag from Variables. See the first template parameter of the instantiation for what Tag is being retrieved and the second template parameter for what Tags are available." 
static_assert(tmpl::list_contains_v<TagList, Tag>,
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/geoffrey/Codes/spectre/spectre/src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:46:14: note: in instantiation of function template

It specifically says it can't retrieve SpacetimeMetric from the Variables. The error then points to the precise line where you try to get it

kidder commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Test_Errors.cpp:
```diff
@@ -0,0 +1,63 @@
+      evolution::Tags::AnalyticCompute<volume_dim, solution_tag,
+                                       solution::tags<DataVector>>,
+      Tags::DifferenceFromAnalyticCompute<system::variables_tag::tags_list,
+                                          solution::tags<DataVector>>>;
```
are these two taglists the same or different?  If the same, please test the subset case...

geoffrey4444 replied:
I chose this test case in part because they are different :)

kidder commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Test_Errors.cpp:
```diff
@@ -0,0 +1,63 @@
+    using tensor_tag = typename decltype(tag_v)::type;
+    const auto& analytic = db::get<Tags::Analytic<tensor_tag>>(box);
+    const auto& diff = db::get<Tags::Difference<tensor_tag>>(box);
+    for (size_t i = 0; i < db::item_type<tensor_tag>::size(); ++i) {
```
do I need this loop? I thought we could do arithmetic with `Variables`

geoffrey4444 replied:
Not when the tags in the variables are different. The analytic Variables has a Tags::Analytic<Tag> for each Tag in the numerical Variables

kidder replied:
never mind, the loop is over a Tensor, not a Variables...

geoffrey4444 replied:
I welcome any suggestions on how to do this with Variables arithmetic, but I couldn't figure out how because the Variables have different tags in them (analytic has Tags::Analytic<Tag>, while numeric has Tag). Is there a way to somehow tell Variables which tags go with which, so the arithmetic works?

kidder replied:
you can't; but why are there two tags lists?  why isn't the AnalyticTagsList constructed from the TagsList?

geoffrey4444 replied:
I tried again, and I still don't see how I can do this arithmetic using Variables, because the tags are different for the lhs and rhs

geoffrey4444 replied:
I'm not sure I understand your comment @kidder...perhaps we could discuss on a hangout if typing isn't getting the job done? 

You can't construct the AnalyticTagsList from TagsList via `db::wrap_tags_in<Tags::Analytic, ...>` because AnalyticTagsList might be a superset of TagsList (e.g., if AnalyticCompute puts in a a Variables with everything the analytic solution provides, which might be more than the evolution variables you want to actually compute differences of).

Also, if the analytic solution had the same TagsList as the numeric solution, then there would be an ambiguity which to retrieve when you try to get one from the DataBox.

kidder commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp:
```diff
@@ -63,6 +63,14 @@ class WrappedGr : public SolutionType {
       typename SolutionType::template tags<DataVector>>;
 
+  template <typename DataType>
+  using tags = tmpl::append<
```
Prefer `tmpl::push_back<LIST, ITEM_0, ITEM_1, ..., ITEM_N>` to `tmpl::append`

geoffrey4444 replied:
Sure thing, but could you please tell me why, so I understand the difference?

kidder replied:
because if you mistakenly do `tmpl::append<LIST,ITEM>` you silently get an empty list

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+ */
+template <typename TagsList, typename AnalyticTagsList>
+struct DifferenceFromAnalyticCompute
+    : db::add_tag_prefix<::Tags::Difference, ::Tags::Variables<TagsList>>,
```
This should be `::Tags::Variables<AnalyticTagsList>`

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+      const ::Variables<db::wrap_tags_in<Tags::Analytic, AnalyticTagsList>>&
+          analytic) noexcept {
+    (*error) = return_type{numeric.number_of_grid_points()};
+    tmpl::for_each<TagsList>([&error, &numeric, &analytic](auto tag_v) {
```
This loop should go over `AnalyticTagsList`

geoffrey4444 replied:
AnalyticTagsList might be a superset of TagsList, so I don't understand how you can loop over AnalyticTagsList here

nilsvu replied:
Now I'm confused. I thought we have a set of system variables in, say, `VariablesTag`. Let `TagsList = db::get_variables_tags_list<VariablesTag>`. Now for a subset of these we have an analytic solution, this subset is `AnalyticTagsList`. In the DataBox we have `db::wrap_tags_in<::Tags::Analytic, AnalyticTagsList>` (may or may not be stored in a Variables). Are we on the same page here?

geoffrey4444 replied:
Let's take the actual example in the test...maybe that will help?

TagsList are the tags we want to return, wrapped in Error<>; these are the quantities we want to compare to the analytic solution. I don't understand why we would only want to compare a subset of these to the analytic solution.

AnalyticCompute puts the analytic solution in the DataBox as a Variables; this is how the analytic solution will be added during a real simulation. But the analytic solutions in GR return many tags besides SpacetimeMetric, Phi, and Pi (namely a bunch of 3+1 quantities), and AnalyticCompute just adds all of these tags. So AnalyticTagsList can be a superset of TagsList.

Does this help? Maybe I'm confused somewhere. I'm happy to discuss further here or in a hangout

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  using base =
+      db::add_tag_prefix<::Tags::Difference, ::Tags::Variables<TagsList>>;
```
Also this should be `AnalyticTagsList`. (for readability I suggest moving this line to the very top of the struct, because it needs to be the same as the struct's base class)

geoffrey4444 replied:
I moved this as you suggest, but I don't understand how this should be AnalyticTagsList, because AnalyticTagsList might be a superset of TagsList

nilsvu commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Test_Errors.cpp:
```diff
@@ -0,0 +1,63 @@
+SPECTRE_TEST_CASE("Unit.AnalyticSolutions.Errors",
+                  "[Unit][PointwiseFunctions]") {
+  static constexpr size_t volume_dim = 1;
+  using system = GeneralizedHarmonic::System<volume_dim>;
```
I suggest to define a few tags for the purpose of this test and make it independent of the GH and evolution code. The code that's being tested only needs a few tags and their `::Tags::Analytic` counterparts in the box, so there's no need to involve the GH system and analytic solutions. You don't need to use the `evolution::Tags::AnalyticCompute` either, just put simple tags in the box.

geoffrey4444 replied:
I thought about this, but I decided that I specifically want to test that this machinery works with with AnalyticCompute and GH. The test you suggest would not have caught the problem with WrappedGr<> that this PR fixes (it didn't have a tags alias), e.g. Also, I want to make sure that if AnalyticCompute ever changes that it won't break DifferenceFromAnalytic, whereas if I add the Analytic<> tags by hand, this might happen

nilsvu replied:
I disagree. This PR implements a unit of functionality that should be tested. If you want to test something with GH and WrappedGr then I suggest you add it to the WrappedGr tests.

geoffrey4444 commented:
Thank you very much for the quick reviews, @kidder and @nilsleiffischer! I pushed fixups addressing your comments, but there's one comment from Larry (about Variables arithmetic) and a few from Nils that I either don't understand or (for the moment at least) am not sure I agree with

geoffrey4444 commented:
@nilsleiffischer and I discussed, and I'll try to make a version that doesn't use Variables at all

nilsvu commented:
@geoffrey4444 here's my suggestion from our call earlier:

```cpp
template <typename TagsList>
struct ErrorsCompute;

template <typename... Tags>
struct ErrorsCompute<tmpl::list<Tags...>>
    : db::add_tag_prefix<::Tags::Error, ::Tags::Variables<tmpl::list<Tags...>>>,
      db::ComputeTag {
  using tags_list = tmpl::list<Tags...>;
  using analytic_tags_list = db::wrap_tags_in<::Tags::Analytic, tags_list>;

  using base = db::add_tag_prefix<::Tags::Error, ::Tags::Variables<tags_list>>;
  using argument_tags = tmpl::append<tags_list, analytic_tags_list>;

  using return_type = tmpl::type_from<base>;
  static void function(
      const gsl::not_null<return_type*> errors,
      const db::const_item_type<Tags>&... fields,
      const db::const_item_type<
          ::Tags::Analytic<Tags>>&... analytic_solutions) noexcept {
    const auto helper = [&errors](const auto error, const auto& field,
                                  const auto& analytic_solution) noexcept {
      for (size_t i = 0; i < field.size(); ++i) {
        (*error)[i] = field[i] - analytic_solution[i];
      }
    };
    EXPAND_PACK_LEFT_TO_RIGHT(
        helper(make_not_null(&get<::Tags::Error<Tags>>(*errors)), fields,
               analytic_solutions));
  }
};
```

kidder commented:
@geoffrey4444 I think the function @nilsleiffischer wrote should take the analytic solution, not tensors for the analytic values, and just compute the analytic values in the function.

geoffrey4444 commented:
> @geoffrey4444 I think the function @nilsleiffischer wrote should take the analytic solution, not tensors for the analytic values, and just compute the analytic values in the function.

Ok, but wouldn’t that duplicate computation of the analytic solution? AnalyticCompute already computes the solution and puts it in the data box during our evolutions...what’s the gain in return for computing the analytic solution again?

Also, that will make this tag take coordinates and time as argument tags...is that ok?

kidder commented:
shouldn't this be the only place the analytic solution is used/computed?   and yes it will need the time and coordinates...

nilsvu commented:
We currently use the analytic solution for imposing boundary conditions and we observe it. The computation would be duplicate for those things. Of course this is only relevant for simulations where we _have_ an analytic solution, which are perhaps not so relevant to optimize.

We have the concept of compute items for the purpose of avoiding such duplicate computations, but currently we’re trying to discourage compute items. I’m unsure what our plan is going forward, so perhaps it’s worth discussing that in a call some time.

Another thing to consider: For elliptic systems I compute the analytic solution only once during initialization, and of course it has no time argument. So if you prefer the version that re-computes the solution for evolution systems, could you still keep the version that uses the Analytic tags from the DataBox for elliptic systems to use?

nilsdeppe commented:
We recompute the analytic solution on the boundary anyway because we in general can't guarantee the volume points and the boundary points line up.

Why should we observe the analytic solution? It seems to me the only things we actually would want to observe are the errors and the numerical solution. Furthermore, ParaView has a nice "Calculator" function that allows you to do basic math, which means given the errors and the numerical solution you can get back the analytic solution if you really want it.

Our plan is to only use compute items for domain/grid related quantities. Everything else will be computed locally in the time derivative computation then discarded. What you want to do in the elliptic code is up to you.

I'm fine keeping a version for elliptic systems. I think we've put too much effort into trying to share code between elliptic and evolution code that doesn't make sense to share. I'd say have a copy in the elliptic directory and a copy in the evolution directory and have them do different things.

geoffrey4444 commented:
> We recompute the analytic solution on the boundary anyway because we in general can't guarantee the volume points and the boundary points line up.
> 
> 
> 
> Why should we observe the analytic solution? It seems to me the only things we actually would want to observe are the errors and the numerical solution. Furthermore, ParaView has a nice "Calculator" function that allows you to do basic math, which means given the errors and the numerical solution you can get back the analytic solution if you really want it.
> 
> 
> 
> Our plan is to only use compute items for domain/grid related quantities. Everything else will be computed locally in the time derivative computation then discarded. What you want to do in the elliptic code is up to you.
> 
> 
> 
> I'm fine keeping a version for elliptic systems. I think we've put too much effort into trying to share code between elliptic and evolution code that doesn't make sense to share. I'd say have a copy in the elliptic directory and a copy in the evolution directory and have them do different things.

Thanks @kidder @nilsdeppe and @nilsleiffischer for helping me see a better way to do this! I’ll go with @nilsleiffischer ’s approach but compute the analytic solution inside the function as @kidder suggests

nilsvu commented:
I agree it's best to have separate code for evolution and elliptic for these things that are just different 👍

geoffrey4444 commented:
@nilsleiffischer @kidder @nilsdeppe I just pushed a fixup that implements your suggestion: 

  * I template on FieldTags... instead of TagsList and AnalyticTagsList (FieldsTags... is not named Tags... because the compiler confuses this with the Tags namespace)
  * I compute the analytic solution internally, instead of getting it from the DataBox. This has the advantage that I can do the subtraction using Variables arithmetic (but I still need a loop over tensor components to load each FieldTag into the errors Variables)
  * I fixed the test failures with separate fixup commits
  * I still prefer testing the ErrorsCompute tag on on a real analytic solution, so that any changes in the analytic solution machinery get caught by the test, to the suggestion of making a dummy analytic solution just for this test

Thanks again for helping me understand a better way to do this while learning a lot about template meta programming and parameter packs! This is ready for another look when you get a chance

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,73 @@
+
+namespace Tags {
+/*!
+ * \brief Prefix indicating the difference between two values, such as betwen a
```
Please adjust the documentation to reflect this now represents the error, not any difference

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,73 @@
+      const db::const_item_type<AnalyticSolutionTag>&
+          analytic_solution_computer,
+      const tnsr::I<DataVector, VolumeDim, Frame::Inertial>& inertial_coords,
+      const double& time, const typename FieldTags::type&... fields) noexcept {
```
- take `double` by value
- `typename FieldTags::type` -> `db::const_item_type<FieldTags>`

geoffrey4444 replied:
I’m fine with these suggestions, but I wonder how I know when to use db::const_item_type or db::item_type vs Tag::type?

nilsvu replied:
Here's my understanding (others can confirm/correct):

- If you know the type explicitly, just write it. E.g. where you know an argument is a `Mesh` or `Scalar<DataVector>`, write that instead of getting the type from some tag.
- If you are dealing with DataBox tags and you're sure they are not base tags, use `db::const_item_type` where the tag is being retrieved from the box and `db::item_type` where it's being mutated.
- If the tags can be base tags, either use `db::[const_]item_type` if you have access to the full DataBox tags list and pass it as second template parameter, or template the function on the argument types and have the compiler infer the types when the function is called.
- If you're sure you're not dealing with a DataBox, or you are sure the tags are simple tags and don't store pointers, you can use `Tag::type`. Prefer `tmpl::type_from<Tag>` over `typename Tag::type` because it's annoying to write `typename`.

geoffrey4444 replied:
Thanks @nilsleiffischer! After discussing with @nilsdeppe, I realize I can avoid getting the type at all by just templating function() on the AnalyticSolution and on the field types...I'll do that in my next fixup

nilsvu replied:
Right, that's the second part of point 3 above. Not sure if (or why) that's preferable to getting the types by `db::const_item_type` when you know they are not base tags. @nilsdeppe would you say it is? Could you confirm or correct my list above?

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,73 @@
+    EXPAND_PACK_LEFT_TO_RIGHT(
+        helper(make_not_null(&get<::Tags::Error<FieldTags>>(*errors)), fields));
+
+    const auto& analytic =
```
delete `&`

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Errors.hpp:
```diff
@@ -0,0 +1,73 @@
+// Distributed under the MIT License.
```
The compute tag is specific to evolution systems. Please move it to `evolution`, I suggest `Evolution/ComputeTags.hpp`. Then you can move the `Errors` prefix tag to `AnalyticSolutions/Tags.hpp`.

geoffrey4444 replied:
@nilsdeppe since you asked me to put this in AnalyticSolutions/Errors.hpp, are you ok with @nilsleiffischer‘s suggestion that I put the compute tag and test back in Evolution/ComputeTags, and move it back into the evolution namespace?

nilsvu reviewed (commented):
From my side please squash the existing fixups as well as changes in response to my comments above.

geoffrey4444 commented:
> From my side please squash the existing fixups as well as changes in response to my comments above.

Thanks! @nilsdeppe and I discussed and agree with your suggestion to move the compute tag back to evolution

geoffrey4444 commented:
@kidder @nilsdeppe would you like to take another look before I squash?

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -40,5 +40,51 @@ struct AnalyticCompute
   }
 };
+
+// @{
```
This should not be necessary. If Doxygen gets confused by the template specialization you can document the forward declaration and `/// \cond` the specialization.

In fact, since you now template the `function` on the `FieldTypes` you don't need the template specialization anymore. You could do this instead:

```cpp
template <size_t VolumeDim, typename AnalyticSolutionTag, typename TagsList>
 struct ErrorsCompute : db::add_tag_prefix<::Tags::Error,
                          ::Tags::Variables<TagsList>>,
       db::ComputeTag {
   using base = db::add_tag_prefix<::Tags::Error, ::Tags::Variables<TagsList>>;
   using return_type = tmpl::type_from<base>;

   using argument_tags = tmpl::append<
       // ...
       TagsList>;

   template <typename AnalyticSolution, typename... ErrorTags, typename... FieldTypes>
   static constexpr void function(
       const gsl::not_null<Variables<tmpl::list<ErrorTags...>>*> errors,
       const AnalyticSolution& analytic_solution_computer,
       const tnsr::I<DataVector, VolumeDim, Frame::Inertial>& inertial_coords,
       const double time, const FieldTypes&... fields) noexcept {
     // ...
     EXPAND_PACK_LEFT_TO_RIGHT(
         helper(make_not_null(&get<ErrorTags>(*errors)), fields));
     // ...
   }
 };
```

nilsvu commented on tests/Unit/Evolution/Test_ComputeTags.cpp:
```diff
@@ -54,3 +58,46 @@ SPECTRE_TEST_CASE("Unit.Evolution.ComputeTags", "[Unit][Evolution]") {
+      tmpl::list<domain::Tags::Coordinates<volume_dim, Frame::Inertial>,
+                 ::Tags::Time, system::variables_tag, solution_tag>;
+  using compute_tags = tmpl::list<evolution::Tags::ErrorsCompute<
+      volume_dim, solution_tag, system::variables_tag::tags_list>>;
```
(not a change request) generally prefer `db::get_variables_tags_list` since that works with prefix tags

I'm not quite sure why this doesn't need a `typename` in front of `system::variables_tag::tags_list` (same in line 85, 88, 95)

nilsdeppe replied:
your question about typename is answered here http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0634r3.html

nilsvu commented on tests/Unit/Evolution/Test_ComputeTags.cpp:
```diff
@@ -54,3 +58,46 @@ SPECTRE_TEST_CASE("Unit.Evolution.ComputeTags", "[Unit][Evolution]") {
+  db::item_type<system::variables_tag> numeric{4, constant_numeric_value};
+
+  const auto box = db::create<db::AddSimpleTags<simple_tags>,
+                              db::AddComputeTags<compute_tags>>(
```
you can just write `db::create<simple_tags, compute_tags>`

geoffrey4444 commented:
Thanks for the comments @nilsleiffischer! My apologies for taking so long to implement them

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -40,5 +40,47 @@ struct AnalyticCompute
   }
 };
+
+// @{
```
You can remove the `// @{}` now

nilsvu reviewed (changes requested):
I suggest you squash all fixups since a lot of code has changed and the fixups are old enough that reviewers don't remember them :)

I still believe the test for the compute tag doesn't need to depend on the intrinsics of the `WrappedGr` solution, but if the other reviewers are OK with that I'll stop bringing it up ;)

geoffrey4444 commented:
> I suggest you squash all fixups since a lot of code has changed and the fixups are old enough that reviewers don't remember them :)
> 
> 
> 
> I still believe the test for the compute tag doesn't need to depend on the intrinsics of the `WrappedGr` solution, but if the other reviewers are OK with that I'll stop bringing it up ;)

Thanks! Nils Deppe asked me not to squash yet, so I’ve been holding off. I prefer the test as is, so I’d like to wait for the other reviewers to weigh in on that question

nilsdeppe reviewed (commented):
LGTM! Sorry for the huge delay. This slipped by me. Please go ahead and squash

geoffrey4444 commented:
Rebased
